File: scratch/test_static_linear_ane.py
# test_static_linear_ane.py — Compiles a Linear layer with static weights for ANE and benchmarks it.
import os
import time
import numpy as np
import torch
import coremltools as ct

class StaticLinearModel(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        # Static weights inside the model
        self.linear = torch.nn.Linear(in_features, out_features, bias=False)
        
    def forward(self, x):
        return self.linear(x)

def main():
    os.makedirs("models", exist_ok=True)
    
    in_features = 2048
    out_features = 2048
    batch_size = 512
    
    print(f"Generating static weight Linear model: input [{batch_size}, {in_features}] x weights [{out_features}, {in_features}]...")
    model = StaticLinearModel(in_features, out_features).eval()
    
    # Input example
    x = torch.randn(batch_size, in_features)
    
    # Trace
    traced = torch.jit.trace(model, x)
    
    # Convert to CoreML forcing ANE
    cml_model = ct.convert(
        traced,
        inputs=[ct.TensorType(name="x", shape=x.shape)],
        compute_units=ct.ComputeUnit.ALL,
        minimum_deployment_target=ct.target.macOS13
    )
    
    out_path = "models/static_linear_2048.mlpackage"
    cml_model.save(out_path)
    print(f"Saved CoreML model to {out_path}")
    
    # Load and run in Python to warm up and verify
    print("Loading model for Python benchmark...")
    loaded_model = ct.models.MLModel(out_path, compute_units=ct.ComputeUnit.CPU_AND_NE)
    
    # Prepare input dict
    x_np = np.random.randn(batch_size, in_features).astype(np.float32)
    inputs = {"x": x_np}
    
    # Warmup
    print("Warming up...")
    for _ in range(5):
        _ = loaded_model.predict(inputs)
        
    # Measure
    print("Benchmarking...")
    times = []
    for _ in range(20):
        t0 = time.perf_counter()
        _ = loaded_model.predict(inputs)
        times.append((time.perf_counter() - t0) * 1000)
        
    lat = np.median(times)
    # 2 * B * I * O
    ops = 2.0 * batch_size * in_features * out_features
    gflops = (ops / lat) / 1_000_000
    tflops = gflops / 1000.0
    
    print("-" * 60)
    print(f"Python CoreML Prediction Latency: {lat:.3f} ms")
    print(f"Throughput: {gflops:.1f} GFLOPS ({tflops:.3f} TFLOPS)")
    print("-" * 60)

if __name__ == "__main__":
    main()

File: scratch/test_static_linear_ane.swift
import Foundation
import CoreML

// Define a simple Tensor-like structure to hold buffer data
class FloatTensor {
    let shape: [Int]
    let count: Int
    let pointer: UnsafeMutablePointer<Float>
    
    init(shape: [Int]) {
        self.shape = shape
        self.count = shape.reduce(1, *)
        self.pointer = UnsafeMutablePointer<Float>.allocate(capacity: count)
        self.pointer.initialize(repeating: 0.0, count: count)
    }
    
    deinit {
        self.pointer.deallocate()
    }
}

func main() {
    print("🚀 Starting Native Swift ANE Linear Benchmark...")
    
    let basePath = "/Users/ommohite/Documents/Programming/FusionML/models"
    let mlpackageURL = URL(fileURLWithPath: "\(basePath)/matmul_2048.mlpackage") // Or static_linear_2048
    let staticLinearURL = URL(fileURLWithPath: "\(basePath)/static_linear_2048.mlpackage")
    
    do {
        print("Compiling CoreML model...")
        let compiledURL = try MLModel.compileModel(at: staticLinearURL)
        print("Compiled URL: \(compiledURL)")
        
        let config = MLModelConfiguration()
        config.computeUnits = .all // Force ANE if possible
        
        print("Loading Model onto ANE...")
        let model = try MLModel(contentsOf: compiledURL, configuration: config)
        
        let B = 512
        let K = 2048
        let O = 2048
        
        print("Creating input MLMultiArray shape [\(B), \(K)]...")
        let xArray = try MLMultiArray(shape: [B, K] as [NSNumber], dataType: .float16)
        
        // Populate input with dummy float16 values
        xArray.withUnsafeMutableBytes { ptr, strides in
            let dest = ptr.baseAddress!.assumingMemoryBound(to: Float16.self)
            for i in 0..<(B * K) {
                dest[i] = Float16(1.0)
            }
        }
        
        let input = try MLDictionaryFeatureProvider(dictionary: [
            "x": MLFeatureValue(multiArray: xArray)
        ])
        
        print("Warming up ANE (5 iterations)...")
        for _ in 0..<5 {
            _ = try model.prediction(from: input)
        }
        
        print("Benchmarking ANE (20 iterations)...")
        var times: [Double] = []
        for _ in 0..<20 {
            let start = CFAbsoluteTimeGetCurrent()
            let output = try model.prediction(from: input)
            let end = CFAbsoluteTimeGetCurrent()
            times.append((end - start) * 1000)
            
            // Touch output to avoid compiler optimization dead-code elimination
            _ = output.featureValue(for: "var_5") // var_5 is the renamed output name in CoreML
        }
        
        times.sort()
        let medianLat = times[times.count / 2]
        let ops = 2.0 * Double(B * K * O)
        let gflops = (ops / medianLat) / 1_000_000
        let tflops = gflops / 1000.0
        
        print("------------------------------------------------------------")
        print(String(format: "  Median ANE Latency: %.3f ms", medianLat))
        print(String(format: "  Throughput:         %.1f GFLOPS (%.3f TFLOPS)", gflops, tflops))
        print("------------------------------------------------------------")
        
    } catch {
        print("❌ Error: \(error)")
    }
}

main()
